### ISSUE TITLE
[PIXELS] Waveform clips when using Int caches

### ISSUE THREAD

**Issue by ThomasWilshaw** [sobotka/olive#52]
**Platform**
Windows 7 SP1 64 Bit
Intel HD Graphics 4600
Olive 0.2.0-6a50c0a6 (Release build)
Built with Visual Studio 2019 and vcpkg libraries

**Summary**
When the cache is set to either 16bit int or 8bit int the waveform appears to clip at about 0.95, even though the pixel sampler is showing 1.0 for both display and reference.

**Additional Information / Output**
Pixel Format 8bit:
![image](https://user-images.githubusercontent.com/7781683/80824858-2698d080-8bd7-11ea-93fe-b25906445804.png)

Pixel Format 16bit float
![image](https://user-images.githubusercontent.com/7781683/80824895-3f08eb00-8bd7-11ea-93be-8b82c55624dd.png)

**Comment by sobotka** [comment]
This depends on the display set, and the fact that the working space to display referred output is accomplished via a single 3D LUT shader in OpenColorIO v1.

If set to a 1D LUT based display, the nature of the edge sampling from the scene referred reference space to the output will never line up perfectly, and as such, it ends up being only an approximation due to performance needs and the use of the GPU. When properly routing through the CPU, or when Olive migrates to OpenColorIO v2, this will be 1:1.

It would also help to not use the sRGB OETF here, and instead use the more-than-likely chance that your actual sRGB display is merely sRGB-like, and uses a native power function.

At some point hopefully soon, Olive will get a better testing configuration.

**Comment by ThomasWilshaw** [comment]
I've also noticed that the waveform doesn't seem to respect changes to the output transforms (Display, Look and View). I'm not sure if that's a related problem though.

**Comment by sobotka** [comment]
Did you right click on the waveform? ;)

**Comment by ThomasWilshaw** [comment]
Haha no. I assumed it would just copy the output view. Is that how it's intended to be long term?

**Comment by sobotka** [comment]
Ask what the value and meaning of a waveform is, and you will likely quickly see why it is important that it reflect a particular transformed data state.

For example, if you are working on that sample footage and you apply a math node to increase exposure via multiply, there might be a want / need to see the result expressed in the camera native Arri Wide Gamut / LogC encoding range to make sure the values are not being clipped.

In another case, it is possible that you’d want to see if your manipulations are being jammed up against the limits of a canonized sRGB encoding, and as such, the waveform would be expected to be indicative of that transform.

It is entirely possible here that your DVL differs from the encoding desired, and as such, it needs to be independent of the particular DVL you are viewing on to be useful. Visualize doing your work on a MacBook with a Display P3 output, which you want to see properly. Yet you are mastering for sRGB, which you want to see in terms of code values and the waveform scope. It becomes clear that the two are separate design needs with different goals. What about dual head? Same issue.

Olive is currently using the DVL transform, but @itsmattkc will expose the waveforms to the full configuration set so this would be possible.

**Comment by itsmattkc** [comment]
@ThomasWilshaw actually appears to be describing a slightly different issue. His waveform very clearly shows 1.0 when using half float.

This issue specifically is because OpenGL clips integer textures to avoid overflow. Naturally this is undesirable as reference buffers are not meant to be clipped. As such it doesn't really make sense for them to be available in Olive and they'll probably be removed soon.

Basically, the solution to this is: don't use integer buffers; they'll be removed soon anyway

(The real question for @sobotka and I is how are you getting 1.0s? Are you using the BT.1886 display?)

**Comment by sobotka** [comment]
@itsmattkc Wow. Looked identical to the edge rabbit hole I pushed you into by mistake. My bad.

I’m sort of in the same boat then; how is the one displaying 1.0 while the other is suspiciously close to the edge sampling error?

**Comment by itsmattkc** [comment]
Ahh I see how it's happening.

Remember the display *will* reach 1.0 if the reference color is > 1.0. It's just reference == 1.0 that gets malformed to 0.94.

So if OP is using a camera log or something that will reach reference > 1.0, then the waveform will indeed come out at 1.0 rather than 0.94.

Unless OP uses integer caches where OpenGL clamps the reference at 0.0 - 1.0, resulting in the all of those 1.0-clamped values transforming into the malformed 0.94 we're all familiar with.

I'm closing this issue in favor of #54.
